Blog post: useR! 2026 — CRDTs, community, and contributing back to R (#345)
posit-dev/open-source-website: content/blog/user-2026-warsaw/index.md

---
title: "useR! 2026: CRDTs, community, and contributing back to R"
date: 2026-07-14
people:
  - Charlie Gao
description: >
  Reflections from useR! 2026 in Warsaw - talking about CRDTs and collaborative editing coming to Quarto 2, increasing adoption of mirai and mori, and a day spent contributing to base R at R Dev Day.
image: "images/hero.jpg"
image-alt: "About two dozen R Dev Day participants gathered for a group photo in a bright, glass-walled atrium, in front of a sign reading 'Enjoy coding at R Dev Day'."
topics:
  - Community
  - Publishing
software:
  - quarto
  - mirai
  - mori
languages:
  - R
tags:
  - useR2026
hidesubscription: false
---

> Although the title of this post starts with "CRDTs" (for maximal alliterative effect), it is in fact the other two (community and contributing back) which are way more important. Sometimes it takes a conference, a substantial chunk of time away from the desk, to step back and re-calibrate our perspectives: to really appreciate the sheer breadth of _good work_ taking place within the community.

Reviewer: ```suggestion
Although the title of this post starts with "CRDTs" (for maximal alliterative effect), it is in fact the other two (community and contributing back) which are way more important to me. Sometimes it takes a conference, a substantial chunk of time away from the desk, to step back and re-calibrate our perspectives: to really appreciate the sheer breadth of _good work_ taking place within the community.
```

You don't need this precise change, but I think you need something in the first sentence to make it clear this is a personal reflection.

And this community is ours. From joining my very first useR! online in 2022, through to Salzburg in 2024 - SatRdays London in between, the very first R Project Sprint in Warwick (UK) in 2023, the R Dev Days organized by Heather Turner and Ella Kaye since, and of course posit::conf(2024) Seattle and posit::conf(2025) Atlanta... I may not have been part of this community for the most number of years, but I've certainly appreciated its warmth and openness.

And so coming back to a European useR! was exciting.

## Distributing state: CRDTs for real-time collaboration

The R community knows how to distribute compute - my talks to date have mainly been centred around this: my creation {[mirai](https://mirai.r-lib.org/)} is an async framework that brings high-performance parallel and distributed computing to R. This lets you send tasks just as easily to a Slurm cluster as to processes on your own machine.

> My talk this time was going to be about something different, but complementary: distributing state. A CRDT is a Conflict-free Replicated Data Type - a data structure that facilitates collaboration by having a property called strong eventual consistency. Without going into detail, this allows it to always merge conflict-free, and is the technology behind most online collaborative editors. Automerge, a particular CRDT implementation, is what we're using in Quarto 2 to make it collaborative out-of-the-box (we're working on this, it's in development).

Reviewer: ```suggestion
My talk this time was about something different, but complementary: distributing state. A CRDT, or Conflict-free Replicated Data Type, is a data structure that facilitates collaboration by having a property called strong eventual consistency. Without going into the details, this allows it to always merge conflict-free, and is the technology behind most online collaborative editors (like Google Docs). Automerge, a particular CRDT implementation, is what we're using in Quarto 2 to make it collaborative out-of-the-box.
```

Author: Google Docs actually doesn't use CRDTs... 😅

> I showed a Shiny app editing a document on one of our Quarto collaborative sync servers. We've created the {[automerge](https://posit-dev.github.io/automerge-r/)} and {[autosync](https://posit-dev.github.io/autosync/)} packages that let R manipulate these structures and sync them over the network - on a par with reference implementations in JavaScript and Rust. Automerge was created by Ink and Switch, an independent research lab for local-first software, and we're glad to provide a link between the two communities.

Reviewer: ```suggestion
In my talk, I showed a Shiny app editing a document on one of our Quarto collaborative sync servers. We've created the {[automerge](https://posit-dev.github.io/automerge-r/)} and {[autosync](https://posit-dev.github.io/autosync/)} packages that let R manipulate these structures and sync them over the network - on a par with reference implementations in JavaScript and Rust. Automerge was created by Ink and Switch, an independent research lab for local-first software, and we're glad to provide a link between the two communities.
```

Reviewer: Missing a segue between the first sentence and the rest of the paragraph.

## Distributing compute: mirai and mori in the wild

> Apart from talking a lot about CRDTs and the three killer features of Quarto 2 (orders of magnitude faster, collaborative editing built-in with live preview, editable in the preview as well as source view), I also got pulled aside by people in the hallways to talk about {[mirai](https://mirai.r-lib.org/)} and {[mori](https://shikokuchuo.net/mori/)}.

Reviewer: ```suggestion
Apart from talking a lot about CRDTs and the three killer features of Quarto 2 (orders of magnitude faster, collaborative editing built-in with live preview, editable in the preview and source views), I also got pulled aside by people in the hallways to talk about {[mirai](https://mirai.r-lib.org/)} and {[mori](https://shikokuchuo.net/mori/)}.
```

Some of these were people working in the life sciences, using these packages for serious science and innovation. This has always been a sector that I've had immense respect for - from my earliest collaboration with Will Landau (Eli Lilly & Co.). Open source software is often regarded as 'high-leverage' in terms of how often and widely it's used, and when used in such high-impact settings, the end goals can be especially motivating.

> It's been extremely gratifying to see that {[mirai](https://mirai.r-lib.org/)} has percolated throughout the community consciousness. Considering that I was only introducing the package at useR! 2024 in Salzburg, it was nice to be able to introduce _myself_ to people this time simply as "the author of mirai".

Reviewer: Needs mori mention too? Maybe something like "Maybe by useR 2027, I'll be able to introduce myself as the author of the mori package"

## The best part: the people

> Given a lot of my colleagues and fellow devs had decided to skip useR! for various reasons, I was slightly apprehensive going in. This fear was dispelled at the very first reception event, where I met so many people from past R Dev Days, amongst them Tina Roszos, who simply beamed at me from across the room! R Core also turned out in force and it was nice to meet Peter Dalgaard and Robert Gentleman for the first time, as well as saying hello again to Luke Tierney, Uwe Ligges and everyone who came.

Reviewer: ```suggestion
Given that only a few of my colleagues could make it to useR! this year, I was slightly apprehensive going in. This fear was dispelled at the very first reception event, where I met so many people from past R Dev Days, amongst them Tina Roszos, who simply beamed at me from across the room! R Core also turned out in force and it was nice to meet Peter Dalgaard and Robert Gentleman for the first time, as well as saying hello again to Luke Tierney, Uwe Ligges and everyone who came.
```
